src/librustc_middle/traits/mod.rs

```diff
@@ -80,8 +81,40 @@ pub enum Reveal {
 }
 
 /// The reason why we incurred this obligation; used for error reporting.
+///
+/// As the happy path does not care about this struct, we store it on the heap
+/// to improve the performance of the happy path.
+///
+/// We do not want to intern this as there are a lot of obligation causes which
+/// only live for a short period of time.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub struct ObligationCause<'tcx> {
+    data: Rc<ObligationCauseData<'tcx>>,
+}
+
+// A dummy obligation. As the parralel compiler does not share `Obligation`s between
+// threads, we use a `thread_local` here so we can keep using an `Rc` inside of `ObligationCause`.
+thread_local! {
+    static DUMMY_OBLIGATION_CAUSE: &'static ObligationCause<'static> = {
+        let boxed = Box::new(ObligationCause::new(DUMMY_SP, hir::CRATE_HIR_ID, MiscObligation));
+        // `thread_local` has the desire to actually execute `Drop::drop` on exit,
+        // which means that a `LocalKey` must not leak references to its content.
+        //
+        // To prevent needless clones and drops, we instead leak this dummy here.
+        Box::leak(boxed)
```

Comment:
This is a memory leak when invoking rustc_driver multiple times inside the same process.